src/main/java/org/traccar/WebDataHandler.java

```diff
@@ -155,32 +181,95 @@ public String formatRequest(Position position) throws UnsupportedEncodingExcepti
+
+            AsyncRequestAndCallback(Position position) {
+
+                String formattedUrl;
+                try {
+                    formattedUrl = json ? url : formatRequest(position);
+                } catch (UnsupportedEncodingException | JsonProcessingException e) {
+                    throw new RuntimeException("Forwarding formatting error", e);
+                }
+
+                requestBuilder = client.target(formattedUrl).request();
+                if (header != null && !header.isEmpty()) {
+                    for (String line: header.split("\\r?\\n")) {
+                        String[] values = line.split(":", 2);
+                        requestBuilder.header(values[0].trim(), values[1].trim());
+                    }
+                }
+
+                if (json) {
+                    payload = prepareJsonPayload(position);
+                }
+
+                deliveryPendingCurrent.incrementAndGet();
+
+                send();
             }
-        }
 
-        Invocation.Builder requestBuilder = client.target(url).request();
+            private void send() {
+                if (json) {
+                    requestBuilder.async().post(Entity.json(payload), this);
+                } else {
+                    requestBuilder.async().get(this);
+                }
+            }
 
-        if (header != null && !header.isEmpty()) {
-            for (String line: header.split("\\r?\\n")) {
-                String[] values = line.split(":", 2);
-                requestBuilder.header(values[0].trim(), values[1].trim());
+            private void retry() {
+                try {
+                    String message = "Position forwarding failed.";
+                    if (!retryEnabled) {
+                        LOGGER.warn(message);
+                    } else {
+                        int pending = deliveryPendingCurrent.get();
+                        if (pending <= deliveryPendingLimit) {
+                            LOGGER.warn(message + " Pending: " + pending
+                                + ". Retrying in " + delay + " seconds.");
+                            GlobalTimer.getTimer().newTimeout(new TimerTask() {
+                                @Override
+                                public void run(Timeout timeout) {
+                                    if (!timeout.isCancelled()) {
+                                        if (delay < retryDelayMax) {
+                                            delay++;
+                                        }
+                                        send();
+                                    }
+                                }
+                            }, delay, TimeUnit.SECONDS);
+                            return;
+                        }
+                        LOGGER.warn(message + " Pending: " + pending
+                            + ". Delivery will not be retried.");
+                    }
+                } catch (Exception e) {
```

Comment:
Why do we need to catch generic exception.

Reply:
The only purpose is to ensure that if forwarding will not be retried (also the case if any exception is thrown), `deliveryPendingCurrent` gets decremented. If you can see a different or better approach, it will be very welcomed.

Reply:
Maybe you can use finally block for anything that needs to be handled.

Reply:
That will be ok, but actually, there's no specific exception I need to handle there. In that case I cannot use a finally block. Do you think I should go that way? Or maybe catch anything, do the job, then re-throw? All of this just to ensure the counter gets decremented when forwarding wont be retried.

Reply:
Sorry, why are you not able to use finally?

Reply:
Sorry again, coming from C++ where we have RAII, I have never used before a try-finally clause (without a catch). Great, now I know it can be done. Thanks.